core/src/main/java/org/apache/iceberg/CherryPickOperation.java

```diff
@@ -0,0 +1,231 @@
+ * <p>
+ * This update is not exposed though the Table API. Instead, it is a package-private part of the Transaction API
+ * intended for use in {@link ManageSnapshots}.
+ */
+class CherryPickOperation extends MergingSnapshotProducer<CherryPickOperation> {
+
+  private final Map<Integer, PartitionSpec> specsById;
+  private Snapshot cherrypickSnapshot = null;
+  private boolean requireFastForward = false;
+  private PartitionSet replacedPartitions = null;
+
+  CherryPickOperation(String tableName, TableOperations ops) {
+    super(tableName, ops);
+    this.specsById = ops.current().specsById();
+  }
+
+  @Override
+  protected CherryPickOperation self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    // snapshotOperation is used by SnapshotProducer when building and writing a new snapshot for cherrypick
+    Preconditions.checkNotNull(cherrypickSnapshot, "[BUG] Detected uninitialized operation");
+    return cherrypickSnapshot.operation();
+  }
+
+  public CherryPickOperation cherrypick(long snapshotId) {
+    TableMetadata current = current();
+    this.cherrypickSnapshot = current.snapshot(snapshotId);
+    ValidationException.check(cherrypickSnapshot != null,
+        "Cannot cherry-pick unknown snapshot ID: %s", snapshotId);
+
+    if (cherrypickSnapshot.operation().equals(DataOperations.APPEND)) {
+      // this property is set on target snapshot that will get published
+      String wapId = WapUtil.validateWapPublish(current, snapshotId);
+      if (wapId != null) {
+        set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+      }
+
+      // link the snapshot about to be published on commit with the picked snapshot
+      set(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, String.valueOf(snapshotId));
+
+      // Pick modifications from the snapshot
+      for (DataFile addedFile : cherrypickSnapshot.addedFiles()) {
+        add(addedFile);
+      }
+
+    } else if (cherrypickSnapshot.operation().equals(DataOperations.OVERWRITE) &&
+        PropertyUtil.propertyAsBoolean(cherrypickSnapshot.summary(), SnapshotSummary.REPLACE_PARTITIONS_PROP, false)) {
+      // the operation was ReplacePartitions. this can be cherry-picked iff the partitions have not been modified.
+      // detecting modification requires finding the new files since the parent was committed, so the parent must be an
+      // ancestor of the current state, or null if the overwrite was based on an empty table.
+      ValidationException.check(
+          cherrypickSnapshot.parentId() == null || isCurrentAncestor(current, cherrypickSnapshot.parentId()),
+          "Cannot cherry-pick overwrite not based on an ancestor of the current state: %s", snapshotId);
+
+      // this property is set on target snapshot that will get published
+      String wapId = WapUtil.validateWapPublish(current, snapshotId);
+      if (wapId != null) {
+        set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
```

Comment:
Will `wapId` stay as a snapshot property or will it eventually become a higher-level tag that references a transaction or branch?

Reply:
Right now this is just separating the existing cherry-pick code into its own class. We haven't completely thought through what changes would be needed to more natively use tags.

More context on this: the way that Adobe uses staged snapshots is to be able to do work in parallel and only commit one staged change even if there are multiple commits with the same wap ID.